br/pkg/lightning/backend/local/local.go

```diff
@@ -1328,28 +1380,74 @@ func (local *local) ImportEngine(ctx context.Context, engineUUID uuid.UUID, regi
 
 	failpoint.Inject("ReadyForImportEngine", func() {})
 
+	err = local.doImport(ctx, lf, regionRanges, regionSplitSize, regionSplitKeys)
+	if err == nil {
+		log.FromContext(ctx).Info("import engine success", zap.Stringer("uuid", engineUUID),
+			zap.Int64("size", lfTotalSize), zap.Int64("kvs", lfLength),
+			zap.Int64("importedSize", lf.importedKVSize.Load()), zap.Int64("importedCount", lf.importedKVCount.Load()))
+	}
+	return err
+}
+
+type firstError struct {
+	err error
+	mu  sync.Mutex
+}
+
+func (e *firstError) save(err error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.err == nil {
+		e.err = err
+	}
+}
+
+func (e *firstError) load() error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	return e.err
+}
+
+func (local *local) doImport(ctx context.Context, engine *Engine, regionRanges []Range, regionSplitSize, regionSplitKeys int64) error {
+	/*
+	   [prepareAndSendJob]-----jobToWorkerCh--->[workers]
```

Comment:
@gozssky Can you give some suggestions on the pattern? Currently regionJob has different stages, multiple workers hold the resouces (pd client, ...) that can advance the stage, and producer/consumer are communicated by channel

Reply:
I can think of two design patterns

one is using goroutines to acquire "worker resource" and then start job, when the job need to sleep some period to retry, release the worker resource. like

```go
for _, j := range jobs {
  w := workerPool.Acquire()
  go func() {j.run(w); w.Release()}
}
```

and

```go
// when meet error
j.w.Release()
sleep()
j.w := workerPool.Acquire()
```

In this method, I want to rely on golang's goroutine scheduling to keep N system thread where N is worker count. But I'm afraid we will spawn too many sleeping goroutines and affect the runtime. 

In the refactoring PR I want to do manually scheduling, where try to keep the N worker goroutines always busy and put the job back when the job needs retry. To solve https://github.com/pingcap/tidb/issues/42456, I should put every time consuming work in worker to make use of concurrency, and remove the barrier in the single-thread producer goroutine. Do you have any suggestions? @gozssky @D3Hunter